### PR TITLE
[mongodb] Collection.find() & .aggregate() cursor generics

### PR DESCRIPTION
**Pull request by hinell**
## Description
Featuring generics on the `Cursor`, `AggregateCursor`, 
`Collection.find()` and `Collection.aggregate()` methods respectively.

Now we can do things like this: 
```ts
interface USER{ _id: ObjectID; name: string;}
let users_col   = db.collection('col_name');
    users.find<USER>()        // Cursor returned
    .forEach(function(currentuser){
       console.log(user.name) // checking USER type
});

interface SUMMARY { id_: ObjectID,  total: number }
let payments = db.collection('payments');
    payments.aggregate<SUMMARY>() // AggregateCursor returned
    .toArray(function(err,sum){
      console.log(sum.total)
  })

```
### Q: Why not to just declare the generic right on the `db.collection()` method and `Collection` interface as proposed by #13731?
Cause there are only a few methods that use the type that generic variable provides and also this would be sort of extra job creating mess and confusing users. So if you feel that you need another one extra generic-typed method like these one then just let me know. 

## Template
- [X] Make your PR against the `master` branch.
- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [X] Run `tsc` without errors.
- [ ] Run `npm run lint package-name` if a `tslint.json` is present.

### CODE REVIEW

**Comment by hinell** [comment]
Attention is needed guys. @CaselIT @beary

**Comment by CaselIT** [comment]
This will break backward compatibility.
As I mentioned https://github.com/DefinitelyTyped/DefinitelyTyped/issues/13731#issuecomment-271079504 I do not think there is a solution that maintains it until the issue https://github.com/Microsoft/TypeScript/issues/2175 is fixed.

If you want to use a typed version manually install it from the typings repo
```
typings i github:Think7/typings-mongodb#generics
```

**Comment by hinell** [comment]
@CaselIT 
>This will break backward compatibility.
You need to explain it more precisely and at what interface.

**Comment by CaselIT** [comment]
If you used any of the classes that are now generic as a type annotation you will not have an error `error TS2314: Generic type '...' requires 1 type argument(s).`

Example if you had:
```ts
let c: Cursor;
```
Typescript will not output `error TS2314: Generic type 'Cursor' requires 1 type argument(s).`

**Comment by hinell** [comment]
>```ts 
>let c: Cursor;
>```
Just specify any type
```ts
let c: Cursor<any>;
```
NO backward compatibility is broken.

**Comment by CaselIT** [comment]
@hinell that is exactly the problem. If you had an existing project that uses the current definition that would break and you would need to go and modify all the classes that are now generics.

**Comment by hinell** [comment]
@CaselIT 
As you can see below travis show us no errros.
And meanwhile whoever need to use the Cursor directly when it is already defined by collection.find() method?

**Comment by CaselIT** [comment]
I'll let the maintainers decide.
I'm not for breaking existing code.

**Comment by hinell** [comment]
@CaselIT It's fine, I need more clear example or description of backward compatibility violation.

**Comment by bowdenk7** [comment]
No response from other authors for over 5 days. Merging.

**Comment by bowdenk7** [comment]
@hinell Looks like there are some merge conflicts. Please refresh the PR so we can merge.

**Comment by hinell** [comment]
@bowdenk7 Here we go...

**Comment by hinell** [comment]
I've tested  both `mongoose` and `lodash` folders by `tsc` and found NO errors. Where the travis do?

File: mongodb/index.d.ts [hunk 1]
```diff
@@ -1222,13 +1226,13 @@ export interface EndCallback {
 
 //http://mongodb.github.io/node-mongodb-native/2.1/api/AggregationCursor.html#~resultCallback
 export type AggregationCursorResult = any | void;
-
+// TODO: change to generic version
```
Comment (aozgaa) [inline mongodb/index.d.ts]: Please remove TODO.

File: mongodb/index.d.ts [hunk 2]
```diff
@@ -1119,24 +1122,25 @@ export interface WriteOpResult {
...
     // http://mongodb.github.io/node-mongodb-native/2.1/api/Cursor.html#addQueryModifier
-    addQueryModifier(name: string, value: boolean): Cursor;
+    addQueryModifier(name: string, value: boolean): Cursor<T>;
     // http://mongodb.github.io/node-mongodb-native/2.1/api/Cursor.html#batchSize
-    batchSize(value: number): Cursor;
+    batchSize(value: number): Cursor<T>;
     // http://mongodb.github.io/node-mongodb-native/2.1/api/Cursor.html#clone
-    clone(): Cursor;
+    // https://github.com/christkv/mongodb-core/blob/2.0/lib/cursor.js#L333
+    clone(): Cursor<T>; // still returns the same type
```
Comment (aozgaa) [inline mongodb/index.d.ts]: Please remove comment.

**Pull request merged**